### PR TITLE
fix: Improve import flight plan from MSFS World Map

### PR DESCRIPTION
**Pull request by frankkopp**
## Summary of Changes
Currently the import from the MSFS World Map is officially supported when the The SimOptions setting "Sync MSFS Flight Plan" is set to `Load Only` or `Save`.

Unfortunately this import is still unreliable and needs imrpovement for users who prefer the MSFS World Map over SimBrief.

Current know issues:

- [x] Order of imported waypoint is wrong
- [x] Waypoints are missing (not imported)
- [x] fix eslint error as far as useful

## Additional context
https://github.com/flybywiresim/docs/pull/691

Test 1:
image
image
`LPPT ROKOB BOMKO ALAMA FTM NEA VRA SSN URUNA IBDER  REPSI  MEBAK  NINTU  MILPA  VADAR  FRI  SUREP  DITON  KUDIS  NUNRI  BETOS  EDDM`

Test 2:
EIDW DE28L D1 DW080 DW682 NEPOD ABIDO PESIT GWC WAFFU ELDAX PIREG SOMIL IRBAL RANUX NEBAX OBAKI BEGAR ELMUR RESIA UNTAD DIKEM ROKIB BABAG PEVAL PUL INTIT MD326 NEVRA KRO LGAV

Discord username (if different from GitHub): Cdr_Maverick#6475

## Testing instructions
- Plan a flight via the MSFS World Map
- Use various combinations of "only enroute", with/without SID, with/without STAR/APPR, etc.
- Test from Gate and Runway (mid air flight plan loading is not yet supported)

OBS: it might be required to enter SID/STAR/APPR in the MCDU again - esp. the STAR/APPR will sometimes not be loaded as expected as the MSFS flight plan does not really handle all arrival and approach combinations (e.g. VIAs).

Tips:
- make screenshots or notes of the planned route from World Map
- compare it to the MCDU F-PLN
- iterate through the MCDU F_PLN on the ND

Updated Docs:
https://github.com/flybywiresim/docs/pull/691

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: src/fmgc/src/flightplanning/FlightPlanAsoboSync.ts [hunk 1]
```diff
@@ -45,142 +45,159 @@ export class FlightPlanAsoboSync {
                 Coherent.call('LOAD_CURRENT_ATC_FLIGHTPLAN').catch(console.error);
                 setTimeout(() => {
                     Coherent.call('GET_FLIGHTPLAN').then(async (data: Record<string, any>) => {
-                        console.log('COHERENT GET_FLIGHTPLAN received');
-                        const { isDirectTo } = data;
+                        console.log('COHERENT GET_FLIGHTPLAN received:');
+                        console.log('Data from MSFS flight plan:', data);
```
Comment (2hwk) [inline src/fmgc/src/flightplanning/FlightPlanAsoboSync.ts:49]: Maybe comment this out for the merge

File: src/fmgc/src/flightplanning/FlightPlanAsoboSync.ts [hunk 2]
```diff
@@ -270,8 +287,8 @@ export class FlightPlanAsoboSync {
                     this.fpChecksum = plan.checksum;
                 }
                 Coherent.call('RECOMPUTE_ACTIVE_WAYPOINT_INDEX')
-                    .catch((e) => console.log('[FP SAVE] Setting Active Waypoint... FAILED'))
-                    .then(() => console.log('[FP SAVE] Setting Active Waypoint... SUCCESS'));
+                    .then(() => console.log('[FP SAVE] Setting Active Waypoint... SUCCESS'))
```
Comment (2hwk) [inline src/fmgc/src/flightplanning/FlightPlanAsoboSync.ts:290]: Comment out debug output

File: src/fmgc/src/flightplanning/ManagedFlightPlan.ts [hunk 3]
```diff
@@ -482,28 +481,21 @@ export class ManagedFlightPlan {
...
-            removed = this.destinationAirfield;
             this.destinationAirfield = undefined;
         } else {
             const segment = this.findSegmentByWaypointIndex(index);
             if (segment) {
-                // console.log("--> REMOVING WAYPOINT ", this.getWaypoint(index), ", FROM SEGMENT ", segment);
                 const spliced = segment.waypoints.splice(index - segment.offset, 1);
-                removed = spliced[0];
-
+                console.log(`removing waypoint ${spliced[0].icao} from segment ${segment.type}`);
```
Comment (2hwk) [inline src/fmgc/src/flightplanning/ManagedFlightPlan.ts:494]: Comment out Debug?

File: src/fmgc/src/flightplanning/ManagedFlightPlan.ts [hunk 4]
```diff
@@ -482,28 +481,21 @@ export class ManagedFlightPlan {
...
                 if (segment.waypoints.length === 0 && segment.type !== SegmentType.Enroute) {
-                    // console.log("SEGMENT LENGTH is 0, REMOVING...");
+                    console.log(`removing segment ${segment.type} as length is 0`);
```
Comment (2hwk) [inline src/fmgc/src/flightplanning/ManagedFlightPlan.ts:496]: Ditto

**Review by 2hwk — CHANGES_REQUESTED** [review]

File: src/fmgc/src/flightplanning/FlightPlanAsoboSync.ts (hunk 1 above)
Comment (frankkopp) [reply src/fmgc/src/flightplanning/FlightPlanAsoboSync.ts:49]: Would like to keep this one to easier debug issue when users report problems.

File: src/fmgc/src/flightplanning/ManagedFlightPlan.ts [hunk 5]
```diff
@@ -200,11 +201,8 @@ export class ManagedFlightPlan {
 
     private computeWaypointEta(distance: number, preComputedTime? :number) {
         const eta = preComputedTime ?? this.computeWaypointTime(distance);
-
         const utcTime = SimVar.GetGlobalVarValue('ZULU TIME', 'seconds');
-
-        // // console.log(`BRUHEGG: ${utcTime}, BRUHHH #2: ${eta}`);
-
+        // console.log(`BRUHEGG: ${utcTime}, BRUHHH #2: ${eta}`);
```
[the remark this answers is not in the document]
Comment (frankkopp) [reply src/fmgc/src/flightplanning/ManagedFlightPlan.ts:205]: :D

File: src/fmgc/src/flightplanning/ManagedFlightPlan.ts (hunk 4 above)
Comment (frankkopp) [reply src/fmgc/src/flightplanning/ManagedFlightPlan.ts:496]: Would like to keep this one to easier debug issue when users report problems.

File: src/fmgc/src/flightplanning/ManagedFlightPlan.ts (hunk 3 above)
Comment (frankkopp) [reply src/fmgc/src/flightplanning/ManagedFlightPlan.ts:494]: Would like to keep this one to easier debug issue when users report problems.

File: src/fmgc/src/flightplanning/FlightPlanAsoboSync.ts (hunk 2 above)
Comment (frankkopp) [reply src/fmgc/src/flightplanning/FlightPlanAsoboSync.ts:290]: Yes - this one is not required

**Review by 2hwk — APPROVED** [review]

**Comment by alexr4339** [comment]
Quality Assurance Tester Report

Discord          : Alex R#6663
Object of testing: #PR 7646
Tier of Testing  : 1
Date             : 22/12/2022

Testing Process: 
Updated the A32NX to latest experimental. I tested 5 different scenarios which I will explain in detail. I want to note, that the import of enroute waypoint worked perfectly everytime.

Test 1: Flight on Worldmap, EDDS to EDDF with SID and STAR/Approach selected on the worldmap and loaded the aircraft on the runway. EDDFs STAR (SPES3C) was inserted twice.
FBW1

Test 2: Flight on Worldmap, EDDS to EDDM with SID and STAR/Approach selected on the worldmap and loaded the aircraft at the gate. Everything imported/loaded as selected on the worldmap.

Test 3: Flight on Worldmap, EDDF to EDDM without SID and without STAR but Approach only (08L) selected on the worldmap and loaded the aircraft on the runway. Everything imported/loaded as selected on the worldmap.

Test 4: Flight on Worldmap, EDDH to EDDF with SID and STAR/Approach (UNOK3C/ILS07C) selected on the worldmap and loaded the aircrafton the runway. The STAR was again inserted twice.
FBW2

Test 5: Flight on Worldmap, EDDF to EDDS with SID and STAR/Approach selected on the worldmap and loaded the aircraft on the runway. Everything imported/loaded as selected on the worldmap.

Negatives:       
Occasional "double-insert" of STAR

Testing Results: 
Passed (As it worked as expected, altough I want to mention I never used the world map import before and thus does not have any reference)

**Comment by frankkopp** [comment]
> Conclusion: I selected pass as the result as the import of enroute waypoints worked and in general the import worked as expected. The double insert of a STAR might be annoying but can be solved by deleting the STAR, so you have the STAR inserted only one time afterwards. The issue of the STAR double-insert did not seem to be a deal-breaker for me.

Thanks for the test!!

There are some issue with procedure still and probably these are not solvable easily - my documentation will state to re-insert procedures in to the MCDU.

But could you give me the exact plans for when it failed? 
Maybe I find something.

**Comment by frankkopp** [comment]
@alexr4339 couldn't reproduce the double STAR insert - tried with what you sent above. 
Do you have the full route info by chance?

**Pull request merged**